Begin integration of changes to use valid transfer sizes

h4yn0nnym0u5e opened the pull request:
- Reduces effect of transmit glitches (usb_audio.cpp:546) 
- ~~Uses 1ms or 500us bInterval depending on bandwidth required (8/96 needs > 1024 bytes / ms, for example)~~
- Tidies up some audio-related macros so they don't have to be put in every build option

This is still experimental, but seems no worse than previous attempts! The transmit sync is obviously still non-functional, so you get the occasional pop. ~~You should simply be able to change sample rate in AudioStream.h and the number of audio channels, and the buffer sizes and bInterval are adjusted to suit~~. Of course that doesn't apply if Full Speed USB is in use, as that's limited to 1k/ms maximum.

ssilverman committed:
Fix HardwareSerial TX and RX FIFO size calculation (teensy4)

mcginty committed:
usb audio: support more than 2 channel inputs

h4yn0nnym0u5e committed:
Re-configure USB descriptors to use AudioStream settings

Kudos to Frank Boesing aka Mcu32 for this!

Also, propagate the USB audio descriptor stuff to all variants that use it (Audio, Serial+MIDI+Audio etc.)

h4yn0nnym0u5e committed:
work-in-progress: Make 8-channel audio output work

* Improve sample count calculation (Will be needed for sample rates other than 44.1kHz)

* Implement higher channel count USB classes
  These are only defined if the settings in usb_desc.h permit, in particular the value of AUDIO_CHANNELS.
  For example, if an attempt is made to use an 8-channel object when only 6 channels are configured, then
  a compile-time error will result.

h4yn0nnym0u5e committed:
Fixes for common sample rate and correct buffer sizes

usb.c - don't assume 44.1kHz
usb_desc.c - use macros for sample size, not fixed values
usb_desc.h - improved rounding for USB buffer sizes

mcginty committed:
Merge pull request #3 from h4yn0nnym0u5e/mcginty/actual-master

Fixes for common sample rate and correct buffer sizes

mcginty committed:
Better documentation for USB descriptor (#4)

better documentation for USB descriptor and usb_audio.*

h4yn0nnym0u5e committed:
Begin integration of changes to use valid transfer sizes

Reduces effect of transmit glitches (usb_audio.cpp:546)
Uses 1ms or 500us bInterval depending on bandwidth required (8/96 needs > 1024 bytes / ms, for example)
Tidies up some audio-related macros so they don't have to be put in every build option

h4yn0nnym0u5e committed:
Begin integration of changes to use valid transfer sizes

Reduces effect of transmit glitches (usb_audio.cpp:546)
Uses 1ms or 500us bInterval depending on bandwidth required (8/96 needs > 1024 bytes / ms, for example)
Tidies up some audio-related macros so they don't have to be put in every build option

mcginty committed:
update feedback algorithm

mcginty committed:
more stable feedback algorithm, 96kHz compatible

mcginty committed:
add floodgate to prevent reporting underflows on power-up

h4yn0nnym0u5e committed:
Fix issue with USB interrupt during audio update

h4yn0nnym0u5e committed:
Fix issue with USB interrupt during audio update

h4yn0nnym0u5e committed:
Fix another IRQ clash; revert to invalid packet sizes

Also remove an unused transmit buffer...

For some reason just halving the buffer size and doubling the frame interval isn't working for input to the Teensy, so for the present we'll test without. Very odd - it seems to do a burst of transfers at 500µs intervals, but then stop, so on average there's only a transfer every millisecond. The data get scrambled, too.

h4yn0nnym0u5e committed:
Fix another IRQ clash; revert to invalid packet sizes

Also remove an unused transmit buffer...

For some reason just halving the buffer size and doubling the frame interval isn't working for input to the Teensy, so for the present we'll test without. Very odd - it seems to do a burst of transfers at 500µs intervals, but then stop, so on average there's only a transfer every millisecond. The data get scrambled, too.

h4yn0nnym0u5e committed:
Merge branch 'PaulStoffregen:master' into mcginty/feature/integrate-01

h4yn0nnym0u5e commented:
I've had to remove the attempt to make buffer sizes valid, as it seems to break something in the USB code resulting in some correct frame timings, but also huge gaps and corrupted data. Very annoying.

However, the latest changes fix some intermittent crashes, and allow use of Audio, Serial_Audio, Serial_MIDI+Audio build options, so they're still useful.

mcginty commented:
I've rebased my `master` against the upstream - if you feel comfortable rebasing your branch now (i.e. `git rebase mcginty/master` if your remote is named `mcginty`) and resolving the conflicts, that'd make the review and merge more straightforward. If not, I can do it when merging after comments.

h4yn0nnym0u5e committed:
Merge branch 'master' of https://github.com/mcginty/teensy-cores into mcginty/actual-master

h4yn0nnym0u5e committed:
Fix up bInterval values

For now this is academic as we haven't got <1ms polling working properly, but assuming we do, then the audio data packets need to be valid sizes with polling intervals to suit. The sync packets are only 4 bytes so guaranteed OK at a fixed interval of 8 microframes.

h4yn0nnym0u5e committed:
Merge branch 'mcginty/feature/integrate-01' of https://github.com/h4yn0nnym0u5e/cores into mcginty/feature/integrate-01

# Resolved Conflicts:
#	teensy4/usb_audio.cpp
#	teensy4/usb_desc.h

h4yn0nnym0u5e commented:
OK, I think I've rebased correctly, without losing anything. I don't have good experiences with rebasing...

I've NOT thoroughly re-tested at this point, as I'm running low on time / energy. Assuming you do test before merging, if there are new issues then just bail and let me know, I'll pick it up when I can ... not expecting you to fix things I've broken!

I have noticed a few things:
- a whole bunch of warnings 
- use of goto (yuck ... but to be fair it does look temporary!)
- debug pin toggling

mcginty commented:
Thanks! No worries about time/energy levels, hope your holidays are going well. Oh yeah, the GOTO is temporarily there for my debug pin timing so I could use my scope to help me understand how the clock drift affected the interrupt timing patterns. There's going to be a good-sized cleanup commit necessary to remove warnings and overly-verbose debug stuff once it's closer to ready for a PR against upstream :).

Maybe the best course then is for me to basically merge what you have as-is and then make modifications on top of it and you can review it asynchronously?

h4yn0nnym0u5e commented:
Sounds good to me! I'll try to keep up to date with your commits on master (or any branches you may create), with a view to checking out stability and timing when using Windows. 

Holiday just starting here for me ... so domestic duties are going to limit hacking time somewhat. Hope you don't get too clobbered by the bomb cyclone - stay safe!

h4yn0nnym0u5e committed:
Add ability to override USB channel count from command line

Needs Extra Stuff to make this useful with Arduino IDE...

h4yn0nnym0u5e committed:
Correct comments

h4yn0nnym0u5e commented:
Having rebased, this is now _not working_ for me. When I send from PC to Teensy (96k, 8 channels, 256-sample blocks), I get no output and eventually a crash which CrashReport says is occurring at the call to report_sink_offset() around line 330 of usb_audio.cpp

This is your code, so I'll let you fix it :-)

mcginty commented:
Was out of town for a few days for the holidays. Will check on that and get back on if I can reproduce that on my side.

mcginty commented on teensy4/usb_desc.h:
```diff
@@ -994,6 +970,55 @@ let me know?  http://forum.pjrc.com/forums/4-Suggestions-amp-Bug-Reports
+	// Change this value to 1024 for correct operation:
+	//                                      VVVVV
+	#define AUDIO_PACKET_SIZE(sz) (((sz) <= 91024)?(sz):((sz)/2))
+	#define AUDIO_INTERVAL(sz)    (((sz) <= 91024)?4:3)
```
I think this operation might need to be rethought. Personally, here's what I was thinking, just based on my use:

* `AUDIO_INTERVAL` is a constant derived from `(AUDIO_FREQUENCY, AUDIO_CHANNELS, AUDIO_SAMPLE_BYTES)` by default, but is also override-able on the CLI just like you did to `AUDIO_CHANNELS` (for my case, for example, I'll be wanting the highest frequency), and then
* `AUDIO_PACKET_SIZE` is a constant derived from `(AUDIO_FREQUENCY, AUDIO_CHANNELS, AUDIO_SAMPLE_BYTES, AUDIO_INTERVAL)`

I'll play with it and try to make a version that looks like it to see how it goes. Let me know if you disagree with the approach.

h4yn0nnym0u5e replied:
That sounds fair. The above was intended as a starting point for keeping the value of AUDIO_PACKET_SIZE() less than or equal to 1024, and only needed two choices for AUDIO_INTERVAL() to get to 96k/8ch/16bit. But a more flexible arrangement is always good...

I believe AUDIO_INTERVAL values have to be between 4 (for a 1ms update) and 1 (for 125µs). I can't see any particular merit in forcing it to a higher frequency than is necessary to get the packet size <=1024 bytes, but there's no harm in having the option. (Technically you can go to bInterval = 16 which is a 4.096s interval! See section 5.6.4 of the USB 2.0 specification).

I had issues with bInterval = 3 / wMaxPacketSize = 776 bytes, which should be about right for 96/8/16. I was getting bursts of USB interrupts at the expected 500µs intervals, but then it'd go quiet so I ended up with an average of an interrupt every 1ms. Possibly a Windows weirdness, maybe related to caching descriptors rather than using the ones just read from the device. You may get better results with Linux.

ssilverman committed:
Make _gettimeofday() weak

The reason is similar to the reason for making _write() weak:
an application may wish to override the behaviour.

ssilverman committed:
Fix build with LTO

Make _serialEvent_default "const" and also make the weak definitions
"extern" (aka "public"). The first change resolves the section type
conflicts and the second change resolves the "weak declaration must be
public" errors.

PaulStoffregen committed:
Merge pull request #671 from ssilverman/make-gettimeofday-weak

Make _gettimeofday() weak

PaulStoffregen committed:
Merge pull request #672 from ssilverman/fix-lto

Fix build with LTO

PaulStoffregen committed:
Add comment with link to explanation of keylayouts defines

PaulStoffregen committed:
Improve printf() support

PaulStoffregen committed:
Print::write(string) gracefully handle null pointer

ssilverman committed:
Change _write() to return Print->write() return value

h4yn0nnym0u5e committed:
Merge branch 'PaulStoffregen:master' into mcginty/feature/integrate-01

PaulStoffregen committed:
Fix IntervalTimer end() when timer interrupt pending
https://forum.pjrc.com/threads/71929-Timer-mystery

PaulStoffregen committed:
Merge pull request #611 from ssilverman/fix-410

Fix HardwareSerial TX and RX FIFO size calculation (teensy4)

PaulStoffregen committed:
Merge pull request #679 from ssilverman/adjust-write-fn

Change _write() to return Print->write() return value

h4yn0nnym0u5e committed:
Merge branch 'PaulStoffregen:master' into mcginty/feature/integrate-01

h4yn0nnym0u5e commented:
Coming up to 3 years old, has been completely ignored, and is now superseded by a working set of changes.
